### ISSUE TITLE
Database save_ensemble method bug

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#271]
I have uncovered an unambiguous bug in the save_ensemble method of Database.  It is another legacy of extending our Ensemble class containers to include an ErrorLogger container along with the ability to kill an entire ensemble with the "kill" method.  

The very first executable line in save_ensemble is this:  
```
       if ensemble_object.dead():
            return
```
That is not the correct behavior.  If the ensemble is marked dead it should first call the method to save any error log messages stored in the ensemble's ErrorLogger container (the symbol name is elog).  

I would fix this, but it isn't quite so trivial.   _save_elog currently handles only atomic TimeSeries or Seismogram objects.  It needs to be extended to ensembles of both atomic types.  The other concern is if this will map correctly into the "tombstone" document we save when atomic data are killed.   If an ensemble is killed I think we will need to create a tombstone document for all ensemble members. A further complication is to be sure the member's error log entries that might have been posted before the ensemble was killed are also saved.   

I think the solution I would recommend is that when save_ensemble gets an ensemble marked dead it will copy the error messages in the ensemble's ErrorLogger container to all members just as we do now for ensemble metadata (the _sync_metadata method of Database), mark all of the ensemble members dead, and then run the rest of save_ensemble_data pretty much as is.

**Comment by pavlis** [comment]
One additional suggestion on this issue.  The save_enemble_data method might want to handle the situation of a completely empty ensemble specially.  The existing logic where such data are simply ignored may be appropriate in that case.

**Comment by haruming** [comment]
That's a good catch. The solution looks good to me. The only concern here is you mention we will need to create a tombstone document for all ensemble members when an ensemble is killed. However, I think our implementation doesn't save atomic ensemble in the database(correct me if I am wrong). Instead, we only save each atomic member(TimeSeries or Seismogram) in the database. Thus, what we can only do is to create or update a tombstone document for each ensemble member, which is handled inside the save_data method.

**Comment by pavlis** [comment]
You are right that we either need to add a separate handler for Ensembles to create a tombstone for ensembles or push clone the errors and post an identical tombstone (other than object id) for a every member.  The later is the easy solution, but has a built in inefficiency.   Saving duplicates of anything, and especially things that might be multiplied by up to thousands (or worse in an iterative procedure killing the same ensemble in multiple passes) is something to worry about.   

I'm aware of the concern about bloated error logs because I just experienced an example.   Debugging my "pwstack" algorithm created an error log with 250,000 documents.  An error in that case, but something that will happen processing large data sets when initial testing misses a pervasive problem.   For example, data problems can surface during an experiment and if you test the workflow on the early data where the problem isn't present you might not realize how bad the problem is.  There is no solution to that problem but we don't want to make it worse by mulitplying error logs by factors of 100 to 1000

**Comment by pavlis** [comment]
Another thought as a postscript to the previous.   Maybe we should implement the easy solution first to correct this bug, but put the more robust solution on our longer range development agenda.

**Comment by wangyinz** [comment]
Yeah, the problem is not only the bloated database, but it is also very bad to the performance. I guess this is not that bad as current Ensemble objects are not well-designed for parallelism anyway. We probably will want to re-design the internal data structure in the future, and that will be the time to completely solve this problem.

**Comment by wangyinz** [comment]
This is resolved by the current ensemble save path. `Database.save_ensemble_data` now delegates to `Database.save_data`, and dead data are passed to `Undertaker.bury`. For an ensemble marked dead, `bury` marks every member dead, persists each member's error log and tombstone, and returns an empty ensemble shell. The entirely-dead-ensemble behavior is also covered in `python/tests/util/test_undertaker.py`. Closing as completed.
